File: Python/2019/2/main.py

def add(a, b, c):
    global I

    I[c] = I[a] + I[b]
    return 4

def mult(a, b, c):
    global I

    I[c] = I[a] * I[b]
    return 4

with open("input.txt") as f:
    I = [int(x) for x in f.read().strip().split(",")]

CI = I[:]
I[1] = 12
I[2] = 2

def process(I):
    index = 0
    
    while I[index] != 99:
        if I[index] == 1:
            index += add(I[index + 1], I[index + 2], I[index + 3])
        elif I[index] == 2:
            index += mult(I[index + 1], I[index + 2], I[index + 3])
        else:
            assert False, I[index]

    return I[0]

print("Part 1:", process(I))

required_output = 19690720

for x in range(99):
    for y in range(99):
        I = CI[:]
        I[1] = x
        I[2] = y

        if process(I) == required_output:
            print("Part 2:", 100 * x + y)
            exit(0)

File: Go/2019/2/main.go
package main

import (
	"fmt"
	"strconv"
	"strings"

	"github.com/TomWatson6/AdventOfCode/Go/2019/intcode"
	"github.com/TomWatson6/AdventOfCode/Go/2019/utils"
)

func main() {
	program, err := readCodes("input.txt")
	if err != nil {
		panic(err)
	}

	programCopy := append([]int{}, program...)

	program[1] = 12
	program[2] = 2

	computer := intcode.New(program)
	err = computer.Run()
	if err != nil {
		panic(err)
	}

	fmt.Printf("Part 1: %d\n", computer.Program[0])

outer:
	for a := 0; a < 100; a++ {
		for b := 0; b < 100; b++ {
			program = append([]int{}, programCopy...)

			program[1] = a
			program[2] = b

			computer = intcode.New(program)
			err = computer.Run()
			if err != nil {
				panic(err)
			}

			if computer.Program[0] == 19690720 {
				fmt.Printf("Part 2: %d\n", a*100+b)
				break outer
			}
		}
	}
}

func readCodes(fileName string) ([]int, error) {
	input, err := utils.ReadFile(fileName)
	if err != nil {
		return nil, err
	}

	parts := strings.Split(input, ",")
	codes := []int{}

	for _, part := range parts {
		code, err := strconv.Atoi(part)
		if err != nil {
			return nil, err
		}

		codes = append(codes, code)
	}

	return codes, nil
}
